receiver/signalfxreceiver/receiver.go

```diff
@@ -0,0 +1,255 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package signalfxreceiver
+
+import (
+	"compress/gzip"
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/gorilla/mux"
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"github.com/open-telemetry/opentelemetry-collector/observability"
+	"github.com/open-telemetry/opentelemetry-collector/oterr"
+	"github.com/open-telemetry/opentelemetry-collector/receiver"
+	"github.com/open-telemetry/opentelemetry-collector/translator/conventions"
+	sfxpb "github.com/signalfx/com_signalfx_metrics_protobuf"
+	"go.opencensus.io/trace"
+	"go.uber.org/zap"
+)
+
+const (
+	defaultServerTimeout = 20 * time.Second
+
+	responseInvalidMethod      = "Only \"POST\" method is supported"
+	responseInvalidContentType = "\"Content-Type\" must be \"application/x-protobuf\""
+	responseInvalidEncoding    = "\"Content-Encoding\" must be \"gzip\" or empty"
+	responseErrGzipReader      = "Error on gzip body"
+	responseErrReadBody        = "Failed to read message body"
+	responseErrUnmarshalBody   = "Failed to unmarshal message body"
+	responseErrNextConsumer    = "Internal Server Error"
+)
+
+var (
+	errNilNextConsumer = errors.New("nil nextConsumer")
+	errEmptyEndpoint   = errors.New("empty endpoint")
+)
+
+// sfxReceiver implements the receiver.TraceReceiver for Zipkin Scribe protocol.
+type sfxReceiver struct {
+	sync.Mutex
+	logger       *zap.Logger
+	config       *Config
+	nextConsumer consumer.MetricsConsumer
+	server       *http.Server
+
+	startOnce sync.Once
+	stopOnce  sync.Once
+}
+
+var _ receiver.MetricsReceiver = (*sfxReceiver)(nil)
+
+func (r *sfxReceiver) MetricsSource() string {
+	const metricsSource string = "SignalFx"
+	return metricsSource
+}
+
+// New creates the SignalFx receiver with the given configuration.
+func New(
+	logger *zap.Logger,
+	config Config,
+	nextConsumer consumer.MetricsConsumer,
+) (receiver.MetricsReceiver, error) {
+
+	if nextConsumer == nil {
+		return nil, errNilNextConsumer
+	}
+
+	if config.Endpoint == "" {
+		return nil, errEmptyEndpoint
+	}
+
+	// Handle config zero values.
+	if config.Name() == "" {
+		config.SetType(typeStr)
+		config.SetName(typeStr)
+	}
```

Comment:
Config type is has public members, in principle they may come with respective values either via factory or the New function. In practice it shouldn't be that relevant but since the name is used for metrics and logging I want to be sure that there is some value.

Reply:
Config loading code always calls SetName after creating the config: https://github.com/open-telemetry/opentelemetry-collector/blob/4e99ac7468fe44c4456acacad28e2fda31beb7ae/config/config.go#L359

Factory's CreateDefaultConfig also set's the name.

This should not be an issue in any of the code paths that exist. It may happen only if `Config` is created manually (which AFAIK we only do in tests) and someone forgets to set the field, but then there is other things that one can forget. I would say if you are going to create `Config` manually then make sure you set the right fields.

To me this code looks suspicious because it makes me think we have a bug elsewhere where it is supposed to be setting the right name when the Config is created. :-)

Reply:
Either way it doesn't matter much, will remove to not raise Qs by people reading the code (I think this is the second time).